examples/textual_inversion/textual_inversion.py

```diff
@@ -825,36 +917,81 @@ def main():
+                noisy_input_images = noise_scheduler.add_noise(input_images, noise, timesteps)
 
                 # Get the text embedding for conditioning
                 encoder_hidden_states = text_encoder(batch["input_ids"])[0].to(dtype=weight_dtype)
 
                 # Predict the noise residual
-                model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
+                model_pred = unet(noisy_input_images, timesteps, encoder_hidden_states).sample
+
+                # https://arxiv.org/abs/2006.11239
+                #
+                # When we predict variance, we use the KL divergence form of the training objective, equation (5).
+                # When variance is fixed, we use the simplified training objective, equation (14).
+                if noise_scheduler.config.variance_type in ["learned", "learned_range"]:
+                    # https://github.com/lucidrains/denoising-diffusion-pytorch/blob/81b4a5de041c0d4d7c19647037388e97cd03f5a8/denoising_diffusion_pytorch/learned_gaussian_diffusion.py#L122
+                    true_mean, _, true_log_variance_clipped = noise_scheduler.q_posterior(
+                        x_start=input_images, x_t=noisy_input_images, t=timesteps
+                    )
+                    model_mean, _, model_log_variance, _ = noise_scheduler.p_mean_variance(
+                        x=noisy_input_images, t=timesteps, model_output=model_pred
+                    )
 
-                # Get the target for loss depending on the prediction type
-                if noise_scheduler.config.prediction_type == "epsilon":
-                    target = noise
-                elif noise_scheduler.config.prediction_type == "v_prediction":
-                    target = noise_scheduler.get_velocity(latents, noise, timesteps)
+                    # kl loss with detached model predicted mean, for stability reasons as in paper
+
+                    detached_model_mean = model_mean.detach()
+
+                    kl = normal_kl(true_mean, true_log_variance_clipped, detached_model_mean, model_log_variance)
+                    kl = meanflat(kl) * NAT
+
+                    decoder_nll = -discretized_gaussian_log_likelihood(
+                        input_images, means=detached_model_mean, log_scales=0.5 * model_log_variance
+                    )
+                    decoder_nll = meanflat(decoder_nll) * NAT
+
+                    # at the first timestep return the decoder NLL, otherwise return KL(q(x_{t-1}|x_t,x_0) || p(x_{t-1}|x_t))
+
+                    vb_losses = torch.where(timesteps == 0, decoder_nll, kl)
+
+                    # simple loss - predicting noise, x0, or x_prev
+
+                    pred_noise, _ = model_pred.chunk(2, dim=1)
+
+                    simple_losses = F.mse_loss(pred_noise, noise)
+
+                    vb_loss_weight = noise_scheduler.config.num_train_timesteps / 1000.0
+
+                    loss = simple_losses + vb_losses.mean() * vb_loss_weight
```

Comment:
If we don't consider the variance we just have normal MSE loss - think that's something worth trying out to begin with. Could we run a first experiment just setting `vb_loss_weight=0` ?